components/utils.py

```diff
@@ -520,3 +606,282 @@ def untar(source, destination='/tmp', strip=1):
     ctx.logger.debug('Extracting {0} to {1}...'.format(source, destination))
     sudo(['tar', '-xzvf', source, '-C', destination,
           '--strip={0}'.format(strip)])
+
+
+def write_file(content, file_path):
+    tmp_file = tempfile.NamedTemporaryFile(delete=False)
+    mkdir(os.path.dirname(os.path.abspath(file_path)))
+    with open(tmp_file.name, 'w') as f:
+        f.write(json.dumps(content))
+    move(tmp_file.name, file_path)
+
+
+class CtxPropertyFactory(object):
+    PROPERTIES_FILE_NAME = 'properties.json'
+    PROPERTIES_PATH = '/opt/cloudify/node_properties/'
+    ROLLBACK_PROPERTIES_PATH = '/opt/cloudify/rollback_node_properties/'
+
+    # A list of property suffixes to be included in the upgrade process,
+    # despite having 'use_existing_on_upgrade' set to ture
+    UPGRADE_PROPS_SUFFIX = ['source_url', 'cloudify_resources_url',
+                            'use_existing_on_upgrade']
+
+    def __init__(self):
+        self.is_upgrade = is_upgrade()
+
+    # Create node properties according to the workflow context install/upgrade
+    def create(self, service_name, write_to_file=True):
```

Comment:
I don't quite understand why we need "Factories" here rather than just regular "Classes". Factories are used to generate objects dynamically based on different parameters whereas here our factories do something completely different than one another.